workspace/codespaces/codespaces.py

```diff
@@ -0,0 +1,88 @@
+import collections
+import datetime
+import json
+import os
+
+import requests
+
+from workspace.utils import blocks
+
+
+URL_PATTERN = "https://api.github.com/orgs/{org}/codespaces"
+TOKEN = os.environ["CODESPACES_GITHUB_API_TOKEN"]
+HEADERS = {
+    "Accept": "application/vnd.github+json",
+    "X-GitHub-Api-Version": "2022-11-28",
+}
+
+
+Codespace = collections.namedtuple(
+    "Codespace", ["age_in_days", "owner", "name", "has_unpushed_changes"]
+)
+
+
+def fetch(url, key):
+    """
+    Recursively fetch all records from the GitHub REST API endpoint given by `url`.
+    `key` is the property under which the records are stored.
+    """
+
+    def set_bearer_auth(request):
+        request.headers["Authorization"] = f"Bearer {TOKEN}"
+        return request
+
+    def fetch_one_page(page_url):
+        response = requests.get(page_url, auth=set_bearer_auth, headers=HEADERS)
+        response.raise_for_status()
+        yield from response.json().get(key)
+        if "next" in response.links:
+            yield from fetch_one_page(response.links["next"]["url"])
+
+    yield from fetch_one_page(url)
+
+
+def get_codespaces_for_org(org):
+    now = datetime.datetime.now(datetime.timezone.utc)
+    records = fetch(URL_PATTERN.format(org=org), "codespaces")
+    for record in records:
+        last_used_at = datetime.datetime.fromisoformat(record["last_used_at"])
+        age_in_days = (now - last_used_at).days
+        owner = record["owner"]["login"]
+        name = record["name"]
+        has_unpushed_changes = record["git_status"]["has_unpushed_changes"] == "true"
+        yield Codespace(age_in_days, owner, name, has_unpushed_changes)
+
+
+def is_at_risk(codespace, threshold_in_days):
+    return codespace.age_in_days >= threshold_in_days
+
+
+def main():
+    org = "opensafely"
+    threshold_in_days = 20
+
+    codespaces = get_codespaces_for_org(org)
+    at_risk_codespaces = sorted(
+        (cs for cs in codespaces if is_at_risk(cs, threshold_in_days)),
+        key=lambda cs: cs.age_in_days,
+        reverse=True,
+    )
+
+    items = [
+        (
+            f"`{cs.owner}` has a Codespace that is {cs.age_in_days} days old "
+            f"({cs.name}). "
+            f"It {'has' if cs.has_unpushed_changes else 'does not have'} unpushed changes.\n\n"
```

Comment:
Why only has_unpushed_changes changes and not has_uncommitted_changes as well? If it's just lack of time, then that's something we can look into adding later.

Reply:
Because I'd not thought carefully enough! Thank you for your comment. I'd not accounted for the fact that a repo can have uncommitted changes _and not_ unpushed changes. I'm going to fetch, and report on, both.